File: AutoEmply/Services/ClaudeClient.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using AutoEmply.Models;
using AutoEmply.Services.Prompts;

namespace AutoEmply.Services;

/// <summary>
/// Anthropic Claude API와 통신하는 HTTP 클라이언트.
///
/// 두 가지 모드를 지원한다:
///   1. GenerateLayoutSpecAsync  - 이미지 → LayoutSpec (픽셀 좌표 기반)
///   2. GenerateFormStructureAsync - 이미지 → FormStructure (비율 기반 논리 구조)
///
/// 공통 흐름: API 호출 → 응답 파싱 → 유효성 검증 → 재시도(최대 N회)
/// </summary>
public sealed class ClaudeClient(
    HttpClient httpClient,
    IConfiguration configuration,
    ILogger<ClaudeClient> logger,
    AiModelState aiModelState)
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true,
        NumberHandling = JsonNumberHandling.AllowReadingFromString
    };

    // ═══════════════════════════════════════════
    //  공개 API
    // ═══════════════════════════════════════════

    /// <summary>이미지를 분석하여 Delphi QuickReport LayoutSpec을 생성한다.</summary>
    public async Task<ClaudeLayoutResult> GenerateLayoutSpecAsync(
        string formName,
        string mediaType,
        string fileBase64,
        ResolvedPromptPreset preset,
        CancellationToken ct,
        bool forceNonEmptyItems = false,
        int emptyObjectRetryLevel = 0)
    {
        var toolSchema = ClaudeToolSchemas.BuildLayoutSpecTool();
        var toolName = "emit_layout_spec";

        var parseResult = await CallClaudeWithRetriesAsync(
            formName, mediaType, fileBase64, preset, toolSchema, toolName, ct,
            // 파싱: 원시 JSON → LayoutSpec
            (rawText, attempt) =>
            {
                if (IsEmptyObjectPayload(rawText) && emptyObjectRetryLevel < 2)
                    return ParseOutcome<LayoutSpec>.RecursiveRetry;

                var layoutSpec = TryParseJson<LayoutSpec>(rawText, out var parseError);
                if (layoutSpec is null)
                    return ParseOutcome<LayoutSpec>.RetryWith(parseError ?? "Return strict JSON only with top-level items array.");

                var validationErrors = LayoutSpecValidator.Validate(formName, layoutSpec);
                if (validationErrors.Count > 0)
                {
                    if (!forceNonEmptyItems && IsItemsEmptyValidation(validationErrors))
                        return ParseOutcome<LayoutSpec>.RecursiveRetry;

                    return ParseOutcome<LayoutSpec>.RetryWith(string.Join(" | ", validationErrors.Take(5)));
                }

                return ParseOutcome<LayoutSpec>.Succeed(layoutSpec);
            });

        // 빈 응답 재시도가 필요한 경우 → 강화된 제약으로 재귀 호출
        if (parseResult.NeedsRecursiveRetry)
        {
            logger.LogWarning("Claude returned empty/items-empty payload. Retrying with stronger constraints. Level={Level}", emptyObjectRetryLevel + 1);
            return await GenerateLayoutSpecAsync(
                formName, mediaType, fileBase64, preset, ct,
                forceNonEmptyItems: true,
                emptyObjectRetryLevel: emptyObjectRetryLevel + 1);
        }

        if (!parseResult.Success)
            return ClaudeLayoutResult.Fail(parseResult.StatusCode, parseResult.Error!, parseResult.Details);

        return ClaudeLayoutResult.Ok(parseResult.Value!);
    }

    /// <summary>이미지를 분석하여 논리적 FormStructure를 추출한다 (Phase 1).</summary>
    public async Task<ClaudeFormStructureResult> GenerateFormStructureAsync(
        string formName,
        string mediaType,
        string fileBase64,
        ResolvedPromptPreset preset,
        CancellationToken ct)
    {
        var toolSchema = ClaudeToolSchemas.BuildFormStructureTool();
        var toolName = "emit_form_structure";

        var parseResult = await CallClaudeWithRetriesAsync(
            formName, mediaType, fileBase64, preset, toolSchema, toolName, ct,
            (rawText, _) =>
            {
                var structure = TryParseJson<FormStructure>(rawText, out var parseError);
                if (structure is null)
                    return ParseOutcome<FormStructure>.RetryWith(parseError ?? "Return valid FormStructure JSON.");

                var validationErrors = FormStructureValidator.Validate(structure);
                if (validationErrors.Count > 0)
                    return ParseOutcome<FormStructure>.RetryWith(string.Join(" | ", validationErrors.Take(5)));

                return ParseOutcome<FormStructure>.Succeed(structure);
            });

        if (!parseResult.Success)
            return ClaudeFormStructureResult.Fail(parseResult.StatusCode, parseResult.Error!, parseResult.Details);

        return ClaudeFormStructureResult.Ok(parseResult.Value!);
    }

    // ═══════════════════════════════════════════
    //  핵심 재시도 루프 (두 모드가 공유)
    // ═══════════════════════════════════════════

    /// <summary>
    /// Claude API 호출 → 응답 파싱 → 실패 시 재시도를 관리하는 공통 루프.
    /// parseAndValidate 콜백으로 모드별 파싱/검증 로직을 주입받는다.
    /// </summary>
    private async Task<ParseOutcome<T>> CallClaudeWithRetriesAsync<T>(
        string formName,
        string mediaType,
        string fileBase64,
        ResolvedPromptPreset preset,
        object toolSchema,
        string toolName,
        CancellationToken ct,
        Func<string, int, ParseOutcome<T>> parseAndValidate)
    {
        var apiKey = Environment.GetEnvironmentVariable("ANTHROPIC_API_KEY");
        if (string.IsNullOrWhiteSpace(apiKey))
            return ParseOutcome<T>.FailWith(400, "API 키 없음");

        var endpoint = configuration["Anthropic:ApiUrl"] ?? "https://api.anthropic.com/v1/messages";
        var maxRetries = Math.Max(1, configuration.GetValue<int?>("Anthropic:MaxRetryAttempts") ?? 3);

        logger.LogInformation("Claude request. Model={Model}, Endpoint={Endpoint}, MaxRetries={Max}",
            preset.Model, endpoint, maxRetries);

        var userPrompt = RenderUserPromptTemplate(preset.UserPromptTemplate, formName);
        var visualBlock = BuildVisualBlock(mediaType, fileBase64);

        ParseOutcome<T>? lastFailure = null;

        for (var attempt = 1; attempt <= maxRetries; attempt++)
        {
            // ── 1. HTTP 요청 ──
            var payload = BuildPayload(preset, toolSchema, toolName, userPrompt, visualBlock);

            using var request = new HttpRequestMessage(HttpMethod.Post, endpoint);
            request.Headers.Add("x-api-key", apiKey);
            request.Headers.Add("anthropic-version", "2023-06-01");
            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
            request.Content = new StringContent(JsonSerializer.Serialize(payload), Encoding.UTF8, "application/json");

            HttpResponseMessage response;
            string body;
            try
            {
                response = await httpClient.SendAsync(request, ct).ConfigureAwait(false);
                body = await response.Content.ReadAsStringAsync(ct).ConfigureAwait(false);
            }
            catch (OperationCanceledException ex) when (!ct.IsCancellationRequested)
            {
                logger.LogError(ex, "Claude timeout. Attempt={Attempt}/{Max}", attempt, maxRetries);
                return ParseOutcome<T>.FailWith(504, "Claude 요청 시간 초과",
                    ["Anthropic API timeout. Increase Anthropic:RequestTimeoutSeconds or lower MaxTokens."]);
            }
            catch (OperationCanceledException)
            {
                return ParseOutcome<T>.FailWith(504, "Claude 요청 취소됨");
            }
            catch (Exception ex) when (ex is HttpRequestException or IOException)
            {
                logger.LogWarning(ex, "Claude network failure. Attempt={Attempt}/{Max}", attempt, maxRetries);
                if (attempt < maxRetries)
                {
                    await ExponentialBackoffAsync(attempt, ct);
                    continue;
                }
                return ParseOutcome<T>.FailWith(503, "Claude 네트워크 오류");
            }

            // ── 2. 응답 처리 ──
            using (response)
            {
                using var responseDoc = TryParseJsonDocument(body);
                var requestId = ExtractRequestId(response, responseDoc);
                aiModelState.Update(ExtractResponseModel(responseDoc));

                // 2a. HTTP 실패
                if (!response.IsSuccessStatusCode)
                {
                    var statusCode = (int)response.StatusCode;
                    logger.LogError("Claude API failed. Attempt={Attempt}/{Max}, Status={Status}, RequestId={Id}",
                        attempt, maxRetries, statusCode, requestId);

                    var msg = ExtractClaudeErrorMessage(responseDoc) ?? "Claude API request failed.";
                    lastFailure = ParseOutcome<T>.FailWith(statusCode, "레이아웃 생성 실패", [$"ClaudeStatus={statusCode}", msg]);

                    if (attempt < maxRetries && IsTransientStatusCode(statusCode))
                    {
                        await ExponentialBackoffAsync(attempt, ct);
                        continue;
                    }
                    return lastFailure;
                }

                // 2b. 모델 텍스트 추출
                if (!TryExtractModelOutput(responseDoc, out var rawText))
                {
                    logger.LogError("Claude response parsing failed. RequestId={Id}, Body={Body}", requestId, body);
                    return ParseOutcome<T>.FailWith(400, "JSON 파싱 실패(모델 응답이 JSON이 아님)");
                }

                logger.LogInformation("Claude output preview. RequestId={Id}, Preview={Preview}",
                    requestId, Truncate(rawText, 2000));

                // 2c. 모드별 파싱 + 검증
                var outcome = parseAndValidate(rawText, attempt);

                if (outcome.NeedsRecursiveRetry)
                    return outcome;

                if (outcome.Success)
                    return outcome;

                // 파싱 실패 → 재시도
                if (attempt < maxRetries)
                {
                    logger.LogWarning("Claude parse/validation failed. Retrying. Attempt={Attempt}/{Max}, Guidance={G}",
                        attempt, maxRetries, outcome.RetryGuidance);
                    continue;
                }

                return ParseOutcome<T>.FailWith(400, outcome.Error ?? "파싱 실패", outcome.Details);
            }
        }

        return lastFailure ?? ParseOutcome<T>.FailWith(500, "레이아웃 생성 실패");
    }

    // ═══════════════════════════════════════════
    //  Payload 빌더
    // ═══════════════════════════════════════════

    private static object BuildPayload(
        ResolvedPromptPreset preset, object toolSchema, string toolName,
        string userPrompt, object visualBlock) => new
    {
        model = preset.Model,
        max_tokens = preset.MaxTokens,
        temperature = preset.Temperature,
        system = preset.SystemPrompt,
        tools = new[] { toolSchema },
        tool_choice = new { type = "tool", name = toolName },
        messages = new object[]
        {
            new
            {
                role = "user",
                content = new object[]
                {
                    new { type = "text", text = userPrompt },
                    visualBlock
                }
            }
        }
    };

    private static object BuildVisualBlock(string mediaType, string fileBase64)
    {
        var blockType = mediaType.Equals("application/pdf", StringComparison.OrdinalIgnoreCase)
            ? "document"
            : "image";

        return new
        {
            type = blockType,
            source = new { type = "base64", media_type = mediaType, data = fileBase64 }
        };
    }

    // ═══════════════════════════════════════════
    //  응답 파싱 헬퍼
    // ═══════════════════════════════════════════

    private JsonDocument? TryParseJsonDocument(string body)
    {
        try { return JsonDocument.Parse(body); }
        catch (JsonException ex)
        {
            logger.LogWarning(ex, "Failed to parse response body as JSON");
            return null;
        }
    }

    /// <summary>Claude 응답에서 tool_use.input 또는 text 블록의 내용을 추출한다.</summary>
    private static bool TryExtractModelOutput(JsonDocument? doc, out string text)
    {
        text = string.Empty;
        if (doc is null) return false;
        if (!doc.RootElement.TryGetProperty("content", out var content) || content.ValueKind != JsonValueKind.Array)
            return false;

        foreach (var block in content.EnumerateArray())
        {
            // tool_use 블록 우선
            if (block.TryGetProperty("type", out var blockType) &&
                string.Equals(blockType.GetString(), "tool_use", StringComparison.OrdinalIgnoreCase) &&
                block.TryGetProperty("input", out var input) &&
                input.ValueKind == JsonValueKind.Object)
            {
                text = input.GetRawText();
                return !string.IsNullOrWhiteSpace(text);
            }

            // text 블록 폴백
            if (block.TryGetProperty("type", out var type) &&
                string.Equals(type.GetString(), "text", StringComparison.OrdinalIgnoreCase) &&
                block.TryGetProperty("text", out var value))
            {
                text = value.GetString() ?? string.Empty;
                return !string.IsNullOrWhiteSpace(text);
            }
        }

        return false;
    }

    private static string? ExtractRequestId(HttpResponseMessage response, JsonDocument? doc)
    {
        if (response.Headers.TryGetValues("request-id", out var ids)) return ids.FirstOrDefault();
        if (response.Headers.TryGetValues("x-request-id", out var xIds)) return xIds.FirstOrDefault();
        if (doc?.RootElement.TryGetProperty("request_id", out var rid) == true) return rid.GetString();
        return null;
    }

    private static string? ExtractResponseModel(JsonDocument? doc) =>
        doc?.RootElement.TryGetProperty("model", out var m) == true ? m.GetString() : null;

    private static string? ExtractClaudeErrorMessage(JsonDocument? doc)
    {
        if (doc?.RootElement.ValueKind != JsonValueKind.Object) return null;
        if (doc.RootElement.TryGetProperty("error", out var err) &&
            err.ValueKind == JsonValueKind.Object &&
            err.TryGetProperty("message", out var msg))
            return msg.GetString();
        return null;
    }

    // ═══════════════════════════════════════════
    //  JSON 역직렬화
    // ═══════════════════════════════════════════

    /// <summary>
    /// 원시 텍스트를 T로 역직렬화한다.
    /// 첫 시도 실패 시, 텍스트에서 JSON 객체 경계({...})를 추출하여 재시도한다.
    /// </summary>
    private static T? TryParseJson<T>(string rawText, out string? parseError) where T : class
    {
        parseError = null;

        var result = Deserialize<T>(rawText, out parseError);
        if (result is not null) return result;

        // 주변 텍스트 제거 후 재시도
        var first = rawText.IndexOf('{');
        var last = rawText.LastIndexOf('}');
        if (first < 0 || last <= first)
        {
            parseError ??= "No JSON object boundaries found in model output.";
            return null;
        }

        return Deserialize<T>(rawText[first..(last + 1)], out parseError);
    }

    private static T? Deserialize<T>(string text, out string? parseError) where T : class
    {
        parseError = null;
        try { return JsonSerializer.Deserialize<T>(text, JsonOptions); }
        catch (JsonException ex) { parseError = $"JSON schema/type mismatch: {ex.Message}"; return null; }
        catch { parseError = "Unknown JSON deserialization error."; return null; }
    }

    // ═══════════════════════════════════════════
    //  유틸리티
    // ═══════════════════════════════════════════

    private static string RenderUserPromptTemplate(string template, string formName) =>
        (template ?? string.Empty)
            .Replace("{{formName}}", formName, StringComparison.Ordinal)
            .Replace("{formName}", formName, StringComparison.Ordinal)
            .Trim();

    private static bool IsTransientStatusCode(int code) =>
        code is 429 or 500 or 502 or 503 or 504 or 529;

    private static bool IsEmptyObjectPayload(string text) =>
        string.Equals(text.Trim(), "{}", StringComparison.Ordinal);

    private static bool IsItemsEmptyValidation(IReadOnlyCollection<string> errors) =>
        errors.Any(e => e.Contains("layoutSpec.items must contain at least one item.", StringComparison.OrdinalIgnoreCase));

    private static string Truncate(string value, int max) =>
        string.IsNullOrEmpty(value) || value.Length <= max ? value : value[..max];

    private static async Task ExponentialBackoffAsync(int attempt, CancellationToken ct) =>
        await Task.Delay((int)Math.Pow(2, attempt - 1) * 1000, ct).ConfigureAwait(false);

    // ═══════════════════════════════════════════
    //  ParseOutcome - 파싱 결과 래퍼
    // ═══════════════════════════════════════════

    /// <summary>
    /// CallClaudeWithRetriesAsync의 parseAndValidate 콜백이 반환하는 결과.
    /// 성공 / 재시도 필요 / 재귀 재시도 필요 / 최종 실패 네 가지 상태를 표현한다.
    /// </summary>
    private sealed record ParseOutcome<T>(
        bool Success,
        int StatusCode,
        string? Error,
        IReadOnlyCollection<string>? Details,
        T? Value,
        string? RetryGuidance = null,
        bool NeedsRecursiveRetry = false)
    {
        public static ParseOutcome<T> Succeed(T value) =>
            new(true, 200, null, null, value);

        public static ParseOutcome<T> FailWith(int statusCode, string error, IReadOnlyCollection<string>? details = null) =>
            new(false, statusCode, error, details, default);

        public static ParseOutcome<T> RetryWith(string guidance) =>
            new(false, 0, guidance, null, default, RetryGuidance: guidance);

        public static ParseOutcome<T> RecursiveRetry =>
            new(false, 0, null, null, default, NeedsRecursiveRetry: true);
    }
}

File: AutoEmply-server/src/main/java/health/autoemplyserver/service/ClaudeClient.java
package health.autoemplyserver.service;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import health.autoemplyserver.config.AiProperties;
import health.autoemplyserver.model.FormStructure;
import health.autoemplyserver.model.LayoutSpec;
import health.autoemplyserver.service.ai.ClaudePayloadFactory;
import health.autoemplyserver.service.ai.ClaudeResponseExtractor;
import health.autoemplyserver.service.prompt.ResolvedPromptPreset;
import health.autoemplyserver.support.exception.BadRequestException;
import health.autoemplyserver.support.exception.ExternalServiceException;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.function.BiFunction;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

@Component
public class ClaudeClient {

    private static final Logger log = LoggerFactory.getLogger(ClaudeClient.class);

    private final ObjectMapper objectMapper;
    private final ClaudeToolSchemas claudeToolSchemas;
    private final LayoutSpecValidator layoutSpecValidator;
    private final FormStructureValidator formStructureValidator;
    private final FormStructurePromptBuilder formStructurePromptBuilder;
    private final ClaudePayloadFactory claudePayloadFactory;
    private final ClaudeResponseExtractor claudeResponseExtractor;
    private final AiModelState aiModelState;
    private final AiProperties aiProperties;
    private final HttpClient httpClient;

    @Autowired
    public ClaudeClient(
        ObjectMapper objectMapper,
        ClaudeToolSchemas claudeToolSchemas,
        LayoutSpecValidator layoutSpecValidator,
        FormStructureValidator formStructureValidator,
        FormStructurePromptBuilder formStructurePromptBuilder,
        ClaudePayloadFactory claudePayloadFactory,
        ClaudeResponseExtractor claudeResponseExtractor,
        AiModelState aiModelState,
        AiProperties aiProperties
    ) {
        this(
            objectMapper,
            claudeToolSchemas,
            layoutSpecValidator,
            formStructureValidator,
            formStructurePromptBuilder,
            claudePayloadFactory,
            claudeResponseExtractor,
            aiModelState,
            aiProperties,
            HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(Math.max(10, aiProperties.getRequestTimeoutSeconds())))
                .build()
        );
    }

    ClaudeClient(
        ObjectMapper objectMapper,
        ClaudeToolSchemas claudeToolSchemas,
        LayoutSpecValidator layoutSpecValidator,
        FormStructureValidator formStructureValidator,
        FormStructurePromptBuilder formStructurePromptBuilder,
        ClaudePayloadFactory claudePayloadFactory,
        ClaudeResponseExtractor claudeResponseExtractor,
        AiModelState aiModelState,
        AiProperties aiProperties,
        HttpClient httpClient
    ) {
        this.objectMapper = objectMapper;
        this.claudeToolSchemas = claudeToolSchemas;
        this.layoutSpecValidator = layoutSpecValidator;
        this.formStructureValidator = formStructureValidator;
        this.formStructurePromptBuilder = formStructurePromptBuilder;
        this.claudePayloadFactory = claudePayloadFactory;
        this.claudeResponseExtractor = claudeResponseExtractor;
        this.aiModelState = aiModelState;
        this.aiProperties = aiProperties;
        this.httpClient = httpClient;
    }

    public LayoutSpec generateLayoutSpec(String formName, String mediaType, String base64Data, ResolvedPromptPreset preset) {
        return generateLayoutSpec(formName, mediaType, base64Data, preset, false, 0);
    }

    private LayoutSpec generateLayoutSpec(String formName, String mediaType, String base64Data, ResolvedPromptPreset preset, boolean forceNonEmptyItems, int emptyRetryLevel) {
        ParseOutcome<LayoutSpec> outcome = callClaudeWithRetries(
            formName,
            mediaType,
            base64Data,
            preset,
            preset.systemPrompt(),
            claudeToolSchemas.buildLayoutSpecTool(),
            "emit_layout_spec",
            appendLayoutGuidance(claudePayloadFactory.renderUserPrompt(preset.userPromptTemplate(), formName), forceNonEmptyItems, null),
            (rawText, attempt) -> {
                if (isEmptyObjectPayload(rawText) && emptyRetryLevel < 2) {
                    return ParseOutcome.recursiveRetry();
                }

                ParseResult<LayoutSpec> parseResult = deserializeToolPayload(rawText, LayoutSpec.class);
                if (!parseResult.success()) {
                    return ParseOutcome.retry(parseResult.message());
                }

                List<String> errors = layoutSpecValidator.validate(formName, parseResult.value());
                if (errors.isEmpty()) {
                    return ParseOutcome.success(parseResult.value());
                }

                boolean itemsEmpty = errors.stream()
                    .anyMatch(error -> error.contains("layoutSpec.items must contain at least one item."));
                if (!forceNonEmptyItems && itemsEmpty && emptyRetryLevel < 2) {
                    return ParseOutcome.recursiveRetry();
                }

                return ParseOutcome.retry(String.join(" | ", errors.stream().limit(5).toList()));
            }
        );

        if (outcome.needsRecursiveRetry()) {
            log.warn("Claude returned empty or items-empty layout payload. Escalating prompt constraints. level={}", emptyRetryLevel + 1);
            return generateLayoutSpec(formName, mediaType, base64Data, preset, true, emptyRetryLevel + 1);
        }

        if (outcome.success()) {
            return outcome.value();
        }

        throw toRuntimeException(outcome);
    }

    public FormStructure generateFormStructure(String formName, String mediaType, String base64Data, ResolvedPromptPreset preset) {
        ParseOutcome<FormStructure> outcome = callClaudeWithRetries(
            formName,
            mediaType,
            base64Data,
            preset,
            formStructurePromptBuilder.buildSystemPrompt(),
            claudeToolSchemas.buildFormStructureTool(),
            "emit_form_structure",
            formStructurePromptBuilder.buildUserPrompt(formName),
            (rawText, attempt) -> {
                ParseResult<FormStructure> parseResult = deserializeToolPayload(rawText, FormStructure.class);
                if (!parseResult.success()) {
                    return ParseOutcome.retry(parseResult.message());
                }

                claudeResponseExtractor.normalizeStructureAliases(parseResult.value());
                List<String> errors = formStructureValidator.validate(parseResult.value());
                if (errors.isEmpty()) {
                    return ParseOutcome.success(parseResult.value());
                }

                return ParseOutcome.retry(String.join(" | ", errors.stream().limit(5).toList()));
            }
        );

        if (outcome.success()) {
            return outcome.value();
        }

        throw toRuntimeException(outcome);
    }

    private <T> ParseOutcome<T> callClaudeWithRetries(
        String formName,
        String mediaType,
        String base64Data,
        ResolvedPromptPreset preset,
        String systemPrompt,
        Map<String, Object> toolSchema,
        String toolName,
        String baseUserPrompt,
        BiFunction<String, Integer, ParseOutcome<T>> parseAndValidate
    ) {
        String apiKey = aiProperties.getApiKey();
        if (apiKey == null || apiKey.isBlank()) {
            return ParseOutcome.failure(400, "ANTHROPIC_API_KEY is required.", List.of());
        }

        int maxRetries = Math.max(1, aiProperties.getMaxRetryAttempts());
        ParseOutcome<T> lastFailure = null;
        String retryGuidance = null;

        for (int attempt = 1; attempt <= maxRetries; attempt++) {
            String userPrompt = appendRetryGuidance(baseUserPrompt, retryGuidance, attempt);
            Map<String, Object> payload = claudePayloadFactory.build(
                preset,
                systemPrompt,
                toolSchema,
                toolName,
                userPrompt,
                mediaType,
                base64Data
            );

            HttpRequest request;
            try {
                request = HttpRequest.newBuilder()
                    .uri(URI.create(aiProperties.getApiUrl()))
                    .timeout(Duration.ofSeconds(Math.max(30, aiProperties.getRequestTimeoutSeconds())))
                    .header("x-api-key", apiKey)
                    .header("anthropic-version", "2023-06-01")
                    .header("content-type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(objectMapper.writeValueAsString(payload)))
                    .build();
            } catch (JsonProcessingException exception) {
                throw new IllegalStateException("Failed to serialize Claude payload.", exception);
            }

            try {
                log.info("Claude request attempt {}/{} model={} mediaType={} formName={}", attempt, maxRetries, preset.model(), mediaType, formName);
                HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
                JsonNode root = safeReadTree(response.body());
                String requestId = extractRequestId(response, root);
                updateResponseModel(root, preset.model());

                if (response.statusCode() < 200 || response.statusCode() >= 300) {
                    String message = claudeResponseExtractor.extractErrorMessage(root);
                    log.warn("Claude API failure attempt={}/{} status={} requestId={} message={}", attempt, maxRetries, response.statusCode(), requestId, message);

                    lastFailure = ParseOutcome.failure(
                        response.statusCode(),
                        "Claude API request failed.",
                        List.of("ClaudeStatus=" + response.statusCode(), message)
                    );

                    if (attempt < maxRetries && isTransient(response.statusCode())) {
                        sleepBackoff(attempt);
                        continue;
                    }
                    return lastFailure;
                }

                String rawText = claudeResponseExtractor.extractModelOutput(root);
                if (rawText == null || rawText.isBlank()) {
                    log.warn("Claude response had no usable tool output attempt={}/{} requestId={}", attempt, maxRetries, requestId);
                    lastFailure = ParseOutcome.failure(400, "Claude response did not contain tool output.", List.of());
                    if (attempt < maxRetries) {
                        retryGuidance = "Return strict JSON only via the requested tool. Do not answer in prose.";
                        sleepBackoff(attempt);
                        continue;
                    }
                    return lastFailure;
                }

                log.info("Claude response success attempt={}/{} requestId={} preview={}", attempt, maxRetries, requestId, truncate(rawText, 600));

                ParseOutcome<T> outcome = parseAndValidate.apply(rawText, attempt);
                if (outcome.needsRecursiveRetry()) {
                    return outcome;
                }
                if (outcome.success()) {
                    return outcome;
                }

                lastFailure = ParseOutcome.failure(
                    400,
                    outcome.error() == null || outcome.error().isBlank() ? "Failed to parse Claude JSON output." : outcome.error(),
                    outcome.details()
                );

                if (attempt < maxRetries) {
                    retryGuidance = outcome.retryGuidance();
                    log.warn("Claude parse/validation failed attempt={}/{} guidance={}", attempt, maxRetries, retryGuidance);
                    sleepBackoff(attempt);
                    continue;
                }

                return lastFailure;
            } catch (IOException exception) {
                log.warn("Claude network failure attempt={}/{}", attempt, maxRetries, exception);
                lastFailure = ParseOutcome.failure(503, "Claude network request failed.", List.of(exception.getMessage()));
                if (attempt < maxRetries) {
                    try {
                        sleepBackoff(attempt);
                        continue;
                    } catch (InterruptedException interruptedException) {
                        Thread.currentThread().interrupt();
                        return ParseOutcome.failure(504, "Claude request interrupted.", List.of(interruptedException.getMessage()));
                    }
                }
                return lastFailure;
            } catch (InterruptedException exception) {
                Thread.currentThread().interrupt();
                return ParseOutcome.failure(504, "Claude request interrupted.", List.of(exception.getMessage()));
            }
        }

        return lastFailure == null
            ? ParseOutcome.failure(502, "Claude request failed.", List.of())
            : lastFailure;
    }

    private RuntimeException toRuntimeException(ParseOutcome<?> outcome) {
        if (outcome.statusCode() >= 500) {
            return new ExternalServiceException(outcome.error() == null ? "Claude request failed." : outcome.error());
        }
        return new BadRequestException(
            outcome.error() == null ? "Failed to parse Claude JSON output." : outcome.error(),
            outcome.details() == null ? List.of() : outcome.details()
        );
    }

    private <T> ParseResult<T> deserializeToolPayload(String rawText, Class<T> type) {
        try {
            return ParseResult.success(objectMapper.readValue(rawText, type));
        } catch (JsonProcessingException exception) {
            int first = rawText.indexOf('{');
            int last = rawText.lastIndexOf('}');
            if (first >= 0 && last > first) {
                try {
                    return ParseResult.success(objectMapper.readValue(rawText.substring(first, last + 1), type));
                } catch (JsonProcessingException ignored) {
                }
            }
            return ParseResult.failure("JSON schema/type mismatch: " + exception.getOriginalMessage());
        }
    }

    private JsonNode safeReadTree(String json) {
        try {
            return objectMapper.readTree(json);
        } catch (JsonProcessingException exception) {
            log.warn("Failed to parse Claude response body as JSON.");
            return null;
        }
    }

    private void updateResponseModel(JsonNode root, String fallbackModel) {
        if (root != null && root.hasNonNull("model")) {
            aiModelState.update(root.get("model").asText());
            return;
        }
        aiModelState.update(fallbackModel);
    }

    private String extractRequestId(HttpResponse<String> response, JsonNode root) {
        String requestId = response.headers().firstValue("request-id").orElse(null);
        if (requestId != null && !requestId.isBlank()) {
            return requestId;
        }

        requestId = response.headers().firstValue("x-request-id").orElse(null);
        if (requestId != null && !requestId.isBlank()) {
            return requestId;
        }

        if (root != null && root.hasNonNull("request_id")) {
            return root.get("request_id").asText();
        }
        return "unknown";
    }

    private boolean isTransient(int statusCode) {
        return statusCode == 429 || statusCode == 500 || statusCode == 502 || statusCode == 503 || statusCode == 504 || statusCode == 529;
    }

    private boolean isEmptyObjectPayload(String rawText) {
        return "{}".equals(rawText == null ? "" : rawText.trim());
    }

    private String appendLayoutGuidance(String prompt, boolean forceNonEmptyItems, String retryGuidance) {
        StringBuilder builder = new StringBuilder(prompt == null ? "" : prompt.trim());
        if (forceNonEmptyItems) {
            builder.append("\n\nHard requirement: return a non-empty items array. Do not return {} or {\"items\":[]}.");
        }
        if (retryGuidance != null && !retryGuidance.isBlank()) {
            builder.append("\n\nRetry requirement: ").append(retryGuidance);
        }
        return builder.toString().trim();
    }

    private String appendRetryGuidance(String prompt, String retryGuidance, int attempt) {
        if (attempt <= 1 || retryGuidance == null || retryGuidance.isBlank()) {
            return prompt;
        }
        return (prompt + "\n\nPrevious output was invalid. Fix this issue and return strict JSON only: " + retryGuidance).trim();
    }

    private String truncate(String value, int max) {
        if (value == null || value.length() <= max) {
            return value;
        }
        return value.substring(0, max);
    }

    private void sleepBackoff(int attempt) throws InterruptedException {
        Thread.sleep((long) Math.pow(2, attempt - 1) * 1000L);
    }

    private record ParseResult<T>(boolean success, T value, String message) {
        static <T> ParseResult<T> success(T value) {
            return new ParseResult<>(true, value, null);
        }

        static <T> ParseResult<T> failure(String message) {
            return new ParseResult<>(false, null, message);
        }
    }

    private record ParseOutcome<T>(
        boolean success,
        int statusCode,
        String error,
        List<String> details,
        T value,
        String retryGuidance,
        boolean needsRecursiveRetry
    ) {
        static <T> ParseOutcome<T> success(T value) {
            return new ParseOutcome<>(true, 200, null, List.of(), value, null, false);
        }

        static <T> ParseOutcome<T> retry(String retryGuidance) {
            return new ParseOutcome<>(false, 0, null, List.of(), null, retryGuidance, false);
        }

        static <T> ParseOutcome<T> recursiveRetry() {
            return new ParseOutcome<>(false, 0, null, List.of(), null, null, true);
        }

        static <T> ParseOutcome<T> failure(int statusCode, String error, List<String> details) {
            return new ParseOutcome<>(false, statusCode, error, details == null ? List.of() : List.copyOf(details), null, null, false);
        }
    }
}
